Mathlib/Algebra/BigOperators/Order.lean

```diff
@@ -816,3 +825,147 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
 #align is_absolute_value.map_prod IsAbsoluteValue.map_prod
 
 end AbsoluteValue
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+
+/-- The `positivity` extension which proves that `∑ i in s, f i` is nonnegative if `f` is, and
+positive if each `f i` is and `s` is nonempty.
+
+Note that this does not do any complicated reasoning. In particular, it does not try to feed in the
+`i ∈ s` hypothesis to local assumptions, and the only ways it can prove `s` is nonempty is if there
+is a local `s.Nonempty` hypothesis or `s = (univ : Finset ι)` and `Nonempty ι` can be synthesized by
+TC inference.
+
+TODO: The following example does not work
+```
+example (s : Finset ℕ) (f : ℕ → ℤ) (hf : ∀ n, 0 ≤ f n) : 0 ≤ s.sum f := by positivity
+```
+because `compareHyp` can't look for assumptions behind binders.
+-/
+@[positivity Finset.sum _ _]
+def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.sum _ $ι $instα $s $f) =>
+    let i : Q($ι) ← mkFreshExprMVarQ q($ι) .syntheticOpaque
+    have body : Q($α) := Expr.betaRev f #[i]
+    let rbody ← core zα pα body
+    -- Try to show that the sum is positive
+    try
+      let .positive pbody := rbody | failure -- Fail if the body is not provably positive
+      -- TODO: If we replace the next line by
+      -- let ps : Q(Finset.Nonempty $s) ← do
+      -- then the type-ascription is ignored. See leanprover/lean4#3126
+      let (ps : Q(Finset.Nonempty $s)) ← do
+        try
+          match s with
+          | ~q(@univ _ $fi) => do
+            let _no ← synthInstanceQ q(Nonempty $ι)
+            pure q(Finset.univ_nonempty (α := $ι))
+          | _ => throwError "`s` is not `univ`"
+        catch _ => do
+          let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s)
+            | failure -- Fail if the set is not provably nonempty
+          pure (.fvar fv)
```

Comment:
I feel like this should be its own definition, since I can imagine there are lots more usecases for checking that a finset is not empty. Maybe even an `aesop` ruleset?

Reply:
I don't really know how to set up an aesop ruleset, so it's a TODO for now.